File: backend/src/main/java/com/storehub/entity/AuditLog.java
package com.storehub.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

/**
 * Append-only audit trail row (Phase 5 spec sections 14-15). No update/delete
 * path is ever exposed by {@code AuditLogRepository}/{@code AuditService} —
 * this entity has no setters (an intentional departure from this codebase's
 * usual @Setter-on-everything convention) so a row, once written, cannot be
 * mutated in place by later code. Never stores passwords/tokens/credentials:
 * {@code oldValue}/{@code newValue} are short human-readable summaries built
 * by the calling service, not raw entity dumps.
 */
@Entity
@Table(name = "audit_logs", indexes = {
        @Index(name = "idx_audit_logs_entity", columnList = "entity_type, entity_id"),
        @Index(name = "idx_audit_logs_timestamp", columnList = "timestamp"),
        @Index(name = "idx_audit_logs_user", columnList = "user_id")
})
@Getter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class AuditLog {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "user_id")
    private Long userId;

    @Column(length = 150)
    private String username;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, columnDefinition = "VARCHAR(20)")
    private AuditAction action;

    @Column(length = 50)
    private String module;

    @Column(name = "entity_type", length = 50)
    private String entityType;

    @Column(name = "entity_id")
    private Long entityId;

    @Column(name = "document_number", length = 30)
    private String documentNumber;

    @Column(name = "old_value", columnDefinition = "TEXT")
    private String oldValue;

    @Column(name = "new_value", columnDefinition = "TEXT")
    private String newValue;

    @Column(nullable = false)
    private LocalDateTime timestamp;

    @Column(name = "ip_address", length = 45)
    private String ipAddress;

    @Column(name = "user_agent", length = 255)
    private String userAgent;

    @Column(name = "financial_year_id")
    private Long financialYearId;

    @Column(columnDefinition = "TEXT")
    private String description;

    @PrePersist
    protected void onCreate() {
        if (this.timestamp == null) {
            this.timestamp = LocalDateTime.now();
        }
    }
}

File: frontend/src/types/auditLog.ts
export type AuditAction =
  | 'CREATE'
  | 'UPDATE'
  | 'POST'
  | 'CANCEL'
  | 'DELETE'
  | 'LOGIN'
  | 'LOGOUT'
  | 'PRINT'
  | 'EXPORT'
  | 'PAYMENT'
  | 'RECEIPT'
  | 'APPROVE'
  | 'REVERSE'
  | 'FY_CLOSE'
  | 'FY_OPEN'
  | 'PERMISSION_CHANGE';

export interface AuditLog {
  id: number;
  userId?: number;
  username?: string;
  action: AuditAction;
  module?: string;
  entityType?: string;
  entityId?: number;
  documentNumber?: string;
  oldValue?: string;
  newValue?: string;
  timestamp: string;
  ipAddress?: string;
  userAgent?: string;
  description?: string;
}
